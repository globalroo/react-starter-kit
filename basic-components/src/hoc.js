import React from "react";

const withStyle = WrappedComponent => {
	return ({ color = "red", ...props }) => (
		<div style={{ backgroundColor: color }}>
			<WrappedComponent {...props} />
		</div>
	);
};

function Button({ label = "default" }) {
	return <button> {label} </button>;
}

export const StylishButton = withStyle(Button);

//Usage
// <StylishButton /> //default
// <StylishButton color="blue" /> // add some color
// <StylishButton label="myButton" /> // add a label
// <StylishButton label="myButton" color="purple"/> //props on styling component and child component */
