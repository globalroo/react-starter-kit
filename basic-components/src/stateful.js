import React, { Component } from "react";

// Usage: <BasicClass />
// If you see this pattern you should probably be using a stateless component
export class BasicClass extends Component {
	render() {
		return <div>Basic Class</div>;
	}
}

// Usage: <BasicClassWithProps message="Basic Class"/>
// If you see this pattern you should probably be using a stateless component
export class BasicClassWithProps extends Component {
	render() {
		return <div>{this.props.message}</div>;
	}
}

// Usage: <BasicClassWithState />
// State should be things that effect the rendering of the component or it's children
// Changing state will trigger a re-render of the component
export class BasicClassWithState extends Component {
	constructor(props) {
		super(props);
		this.state = {
			message: "Basic Class"
		};
	}
	render() {
		return <div>{this.state.message}</div>;
	}
}

// Usage: <BasicClassWithStateUpdate />
export class BasicClassWithStateUpdate extends Component {
	constructor(props) {
		super(props);
		this.state = {
			message: "Basic Class"
		};
	}
	render() {
		return (
			<button
				onClick={() =>
					this.setState({
						message: "State Changed, Component Re-Rendered."
					})
				}
			>
				{this.state.message}
			</button>
		);
	}
}

// A containing component with classes
// Usage:
//	<BasicWithChildren>
//		<p>I'm a child of the Component</p>
//	</BasicWithChildren>
export class BasicWithChildren extends Component {
	render() {
		return <div>{this.props.children}</div>;
	}
}
