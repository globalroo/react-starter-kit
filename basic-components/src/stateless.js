// Needed for JSX support
import React from "react";

// File uses function expressions
// Can just as easily use: export function Basic() { return <div>Basic Component</div> }

// Usage: <Basic />
export const Basic = () => <div>Basic Component</div>;

// Usage: <BasicWithProps message="Hello World!" /> - destructure properties to make life easier
export const BasicWithProps = ({ message }) => <div>{message}</div>;

// Usage: <BasicWithDOMAttrs /> Some attributes require camelCase
// https://reactjs.org/docs/dom-elements.html
export const BasicWithDOMAttrs = () => (
	<div>
		<input
			className="my-class"
			data-qa="qa-hook"
			readOnly={true}
			tabIndex="1"
		/>
	</div>
);

// Usage: <BasicWithEventHandler /> note the events are camelCase onClick, onChange, onKeyDown
// https://reactjs.org/docs/handling-events.html
export const BasicWithEventHandler = () => (
	<div>
		<button onClick={() => alert("Clicked")}>Click Me</button>
	</div>
);

// Usage: <BasicWithEventSentToParent handleClick={(e) => alert("Child component clicked")} />
export const BasicWithEventSentToParent = ({ handleClick }) => (
	<div>
		<button onClick={handleClick}>Click Me</button>
	</div>
);

// Usage: <BasicWithValuePassedToParent handleClick={(message) => alert(message)} />
export const BasicWithValuePassedToParent = ({ handleClick }) => (
	<div>
		<button onClick={e => handleClick("Clicked")}>Click Me</button>
	</div>
);

// A containing component
// Usage:
//	<BasicWithChildren>
//		<p>I'm a child of the Component</p>
//	</BasicWithChildren>
export const BasicWithChildren = ({ children }) => <div>{children}</div>;
