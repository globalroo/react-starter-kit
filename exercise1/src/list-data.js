export const listData = [
	{
		id: "9e8034d1-8bc2-4d02-b297-bc37138ea923",
		name: "Employee1",
		team: "Team1"
	},
	{
		id: "0efbd2c5-4cd7-4cff-8326-727e7d814c24",
		name: "Employee2",
		team: "Team1"
	},
	{
		id: "7bf24b4b-4df6-4292-a980-78ff0b7bdab3",
		name: "Employee3",
		team: "Team1"
	},
	{
		id: "20d8ac5f-326f-4c31-9078-fe62f7f61c2b",
		name: "Employee4",
		position: "Team Lead",
		team: "Team1"
	},
	{
		id: "603dc941-432b-4911-af78-aed47e60a0b5",
		name: "Employee5",
		team: "Team1"
	},
	{
		id: "72087225-d0b5-4f11-b813-7f2502be5641",
		name: "Employee6",
		team: "Team2"
	},
	{
		id: "abe64748-a1f4-4aba-b3d1-78829cbb071f",
		name: "Employee7",
		team: "Team2"
	},
	{
		id: "473c7c43-3056-48af-a74a-8bc17748b0ee",
		name: "Employee8",
		team: "Team2"
	},
	{
		id: "06e115c5-db80-403e-921d-fcf5b069447c",
		name: "Employee9",
		team: "Team2"
	},
	{
		id: "e2497dd6-19e4-4658-b133-12d5d1b4cb93",
		name: "Employee10",
		team: "Team2"
	},
	{
		id: "ba0b6b64-f13a-45e5-b1e8-dca5cc523c34",
		name: "Employee11",
		position: "Team Lead",
		team: "Team2"
	},
	{
		id: "40033bee-20fb-498a-a237-660ef5413241",
		name: "Employee12",
		team: "Team3"
	},
	{
		id: "37721481-b28b-4e96-b6d6-da6358837b92",
		name: "Employee13",
		team: "Team3"
	},
	{
		id: "231b21ca-7cfa-4772-bec7-fab5c62bea93",
		name: "Employee14",
		team: "Team3"
	},
	{
		id: "3b8f2ad4-8142-4ad4-96d0-76fd2e77d448",
		name: "Employee15",
		team: "Team3"
	},
	{
		id: "74126de4-bf36-4e64-acd2-96f42736a3c0",
		name: "Employee16",
		position: "Team Lead",
		team: "Team3"
	}
];
