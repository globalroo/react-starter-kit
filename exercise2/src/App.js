// Conditional rendering example, calling out to endpoint
// Basic GraphQL introduction with server, GraphiQL and adjustable client query

import React, { Component, Fragment } from "react";
import { getQueryFromServer } from "./graphql-service";

// import resources like this to use in your component
// Alternatively you can use a url on the component or <link> via index.html for css as per usual
import logo from "./logo.svg";
import "./App.css";

// Figure out the required data shape in GraphiQL (localhost:4000/graphql)
const getVideosQuery = "{ videos { key, title, synopsis }}";

// TODO: Task - Make this work too
// TODO - Task (bonus points) - Display a list of related Videos to the selected Video
//const getSpecificVideoQuery = "{ video(id:2) { key, title }}";

const myGraphQL = getVideosQuery;

// TODO: Task - Move these to their own files - make them components, add some style.
const Video = ({ video = {} }) => <li>{JSON.stringify(video)}</li>;

const Videos = ({ videos = [] }) => (
	// TODO: Task - Output the movies as a list of poster images with title below
	<ul>
		{videos.map(({ key, ...videoMeta }) => (
			<li key={key}>{JSON.stringify(videoMeta)}</li>
		))}
	</ul>
);

class App extends Component {
	constructor(props) {
		super(props);

		// Changing this initial state via setState triggers a re-render of the component
		this.state = {
			errors: [],
			loaded: false,
			video: null,
			videos: []
		};
	}

	async componentDidMount() {
		// This is the lifecycle method to use to get data from an external service
		// https://reactjs.org/docs/state-and-lifecycle.html
		const result = await getQueryFromServer(myGraphQL);
		// Update state and trigger re-render
		this.setState({ ...result });
	}

	render() {
		// Conditional Rendering example - Conditional return
		if (this.state.errors.length > 0) {
			// TODO: Task - Make Error handling component that looks nice
			return <h3>GraphQL Error: {JSON.stringify(this.state.errors)}</h3>;
		}

		// TODO: Task - Identify and create component optimisations below.
		return (
			<div className="App">
				<header className="App-header">
					<img src={logo} className="App-logo" alt="GraphQL logo" />
					<p>
						Exercise 2 - <code>Using Endpoints</code>
					</p>
				</header>
				<div className="App-main">
					{/* Conditional rendering inline contrived example - Ternary */}
					{this.state.loaded ? (
						<>
							{/* A Fragment allows adjacent / sibling components to be returned without a parent */}
							{/* Conditional rendering inline contrived example - && */}
							{this.state.video !== null && <Video video={this.state.video} />}
							{this.state.videos.length > 0 && (
								<Videos videos={this.state.videos} />
							)}
						</>
					) : (
						<div>Loading...</div>
					)}
				</div>
			</div>
		);
	}
}

export default App;
