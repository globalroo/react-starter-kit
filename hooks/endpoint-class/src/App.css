.App {
	text-align: center;
}

.App-logo {
	animation: App-logo-spin infinite 20s linear;
	height: 10vmin;
}

.App-body {
	background-color: #242525;
	min-height: 100vh;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	font-size: calc(10px + 2vmin);
	color: white;
}

.App-link {
	color: #61dafb;
}

@keyframes App-logo-spin {
	from {
		transform: rotate(0deg);
	}
	to {
		transform: rotate(360deg);
	}
}

.center-center,
.loader {
	display: -webkit-box;
	display: -moz-box;
	display: -webkit-flex;
	display: -ms-flexbox;
	display: flex;
	-webkit-box-align: center;
	-moz-box-align: center;
	-o-box-align: center;
	-ms-flex-align: center;
	-webkit-align-items: center;
	align-items: center;
	-webkit-box-pack: center;
	-moz-box-pack: center;
	-o-box-pack: center;
	-ms-flex-pack: center;
	-webkit-justify-content: center;
	justify-content: center;
}

.loadingSpinner {
	height: 400px;
}

.App-main {
	font-size: 26px;
	/* font-family: "Press Start 2P"; */
}

.App-main li {
	list-style: none;
}
