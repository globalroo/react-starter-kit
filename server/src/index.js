import express from "express";
import graphqlHTTP from "express-graphql";

import { GraphQLSchema } from "graphql";
import { RootQueryType } from "./types/root-query";

const PORT = process.env.GQLPORT || 4000;
const cors = require("cors");
const app = express();
const delay = require("express-delay");

app.use(cors());

app.use(delay(1000));

const schema = new GraphQLSchema({
	query: RootQueryType
});

app.use(
	"/",
	graphqlHTTP({
		schema,
		graphiql: true
	})
);

const server = app.listen(PORT, () => {
	console.log(`listening on port:${PORT}`);
});

module.exports = server;
