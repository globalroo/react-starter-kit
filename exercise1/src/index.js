import React, { Component } from "react";
import ReactDOM from "react-dom";

import "./styles.css";

// TODO: Style components
// TODO: Add Components to own file
import { listData } from "./list-data";

const ListItem = ({ item }) => {
	const { id, ...useful } = item;
	return <li>{JSON.stringify(useful)}</li>;
};

const List = ({ list }) => (
	<ul id={"the-list"}>
		{list.map(item => {
			return <ListItem key={item.id} item={item} />;
		})}
	</ul>
);
export class ListControl extends Component {
	constructor(props) {
		super(props);
		this.state = {
			searchTerm: "",
			filteredData: [...this.props.listData]
		};
	}

	//TODO - Add a method to filter for each of the squads and add buttons to call them
	teamLeads = () => {
		const teamLeads = this.props.listData.filter(item => item.position);
		this.setState({ filteredData: teamLeads });
	};

	reset = () => this.setState({ filteredData: [...this.props.listData] });

	getSearchTerm = event => event.currentTarget.value;

	getFilteredData(searchTerm = "") {
		return this.props.listData.filter(item =>
			item.name.toLowerCase().includes(searchTerm.toLowerCase())
		);
	}

	unControlledChange = event => {
		const searchTerm = this.getSearchTerm(event);
		const filteredData = this.getFilteredData(searchTerm);
		this.setState(() => ({ filteredData }));
	};

	controlledChange = event => {
		const searchTerm = this.getSearchTerm(event);
		const filteredData = this.getFilteredData(searchTerm);
		this.setState(() => ({ searchTerm, filteredData }));
	};

	domNodeValueAcquisition = () => {
		const searchTerm = this.domNode.value;
		const filteredData = this.getFilteredData(searchTerm);
		this.setState(() => ({ filteredData }));
	};

	handleFormSubmit = event => {
		event.preventDefault();
		const { elements } = event.currentTarget;
		const { name, surname } = elements;
		alert(
			`The value of the forms textfields are ${name.value} and ${surname.value}`
		);
		const filteredData = this.getFilteredData(name.value);
		this.setState(() => ({ filteredData }));
	};

	render() {
		// TODO: Add a button to filter for each of the groups
		// TODO: Can you see patterns in the components? Can you optimise them?
		// Bonus points:
		// TODO: Add a 'FIRED!' Button to remove someone from the list
		// TODO: Add a 'HIRED! Form to add a new record
		return (
			<div>
				<h3>Filter list using state</h3>
				<button onClick={this.teamLeads}> Team Leads </button>
				<button onClick={this.reset}> Reset List </button>
				<hr />
				<h3>Searching with inputs</h3>
				{/* Uncontrolled component acts on event - we just use the value in the component - transient state*/}
				<input
					name="byEvent"
					onChange={this.unControlledChange}
					placeholder="Search using event"
				/>
				<br />
				{/* Controlled component - we're going to send this value elsewhere, into another component or as part of a form - persistent state*/}
				<input
					name="byState"
					onChange={this.controlledChange}
					placeholder="Search using state"
					value={this.state.searchTerm}
				/>
				<br />
				{/* 
					DOM Reference - Get handle on DOM node and manipulate directly. 
					Not particularly React-y but useful for passing nodes to third party libraries 
					It's rare you need to operate on the DOM directly.
				*/}
				<input
					name="byRef"
					placeholder="Search using DOM ->"
					ref={ref => (this.domNode = ref)}
				/>
				<button onClick={this.domNodeValueAcquisition}>Search</button>
				<hr />
				{/* 
					Form Example - by handling onSubmit you can capture all of the values at once if you prefer.
				*/}
				<h3>Working with form data</h3>
				<form onSubmit={this.handleFormSubmit}>
					<input name="name" placeholder="name" />
					<input name="surname" placeholder="surname" />
					<button type="submit">Submit</button>
				</form>
				<hr />
				<List list={this.state.filteredData} />
			</div>
		);
	}
}

function App() {
	return (
		<div className="App">
			<ListControl listData={listData} />
		</div>
	);
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
