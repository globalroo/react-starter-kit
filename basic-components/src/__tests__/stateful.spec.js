import React from "react";

import { shallow, render, mount } from "src/enzyme";
import {
	BasicClass,
	BasicClassWithProps,
	BasicClassWithState
} from "src/stateful";

// Using Enzyme (AirBnB) general ethos is to test in isolation using 'shallow'

describe("Stateful tests", () => {
	it("Renders a basic component and verifies Basic Class is displayed", () => {
		const wrapper = shallow(<BasicClass />);
		expect(wrapper.contains("Basic Class")).toBeTruthy();
	});

	it("Clicks a button and calls back to a method", () => {
		const wrapper = shallow(<BasicClassWithProps message="Basic Class" />);
		expect(wrapper.contains("Basic Class")).toBeTruthy();
	});

	it("Clicks a button and calls back to a method", () => {
		const wrapper = shallow(<BasicClassWithState message="Basic Class" />);
		expect(wrapper).toMatchSnapshot();
	});
});
