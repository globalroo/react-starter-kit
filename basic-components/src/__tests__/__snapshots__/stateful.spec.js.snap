// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stateful tests Clicks a button and calls back to a method 1`] = `
<div>
  Basic Class
</div>
`;
